<a href="https://colab.research.google.com/github/EduardoJBenavidesC/Proyecto-Completo/blob/main/Copia_de_AluraStoreLatam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Importación de datos



In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_1%20.csv"
url2 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_2.csv"
url3 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_3.csv"
url4 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_4.csv"

tienda = pd.read_csv(url)
tienda2 = pd.read_csv(url2)
tienda3 = pd.read_csv(url3)
tienda4 = pd.read_csv(url4)

tienda.head()

Producto Categoría del Producto    Precio  Costo de envío  \
0  Asistente virtual           Electrónicos  164300.0          6900.0   
1    Mesa de comedor                Muebles  192300.0          8400.0   
2      Juego de mesa               Juguetes  209600.0         15900.0   
3         Microondas      Electrodomésticos  757500.0         41000.0   
4   Silla de oficina                Muebles  335200.0         20200.0   

  Fecha de Compra         Vendedor Lugar de Compra  Calificación  \
0      16/01/2021      Pedro Gomez          Bogotá             4   
1      18/05/2022  Beatriz Morales        Medellín             1   
2      15/03/2021   Juan Fernandez       Cartagena             1   
3      03/05/2022   Juan Fernandez            Cali             4   
4      07/11/2020    Maria Alfonso        Medellín             5   

       Método de pago  Cantidad de cuotas       lat       lon  
0  Tarjeta de crédito                   8   4.60971 -74.08175  
1  Tarjeta de crédito                   4   6.25184 -75.56359  
2  Tarjeta de crédito                   1  10.39972 -75.51444  
3               Nequi                   1   3.43722 -76.52250  
4               Nequi                   1   6.25184 -75.56359

# CONOCIENDO LOS DATOS

##1.- Dimensiones de cada base de datos, por tienda.

#### Se hará una tabla que resuma las dimensiones de cada base y así determinar si hay **representatividad estadística** de la información de cada tienda para el momento de sacar conclusiones; y a su vez, visualizar de manera más ordenada y simple la estructura de la información.

In [23]:
import pandas as pd

# Definición de colecciones maestras
las_tiendas = ['tienda', 'tienda2', 'tienda3', 'tienda4']
bases = [tienda, tienda2, tienda3, tienda4]

resumen_datos = []
for nombre_tienda, df in zip(las_tiendas, bases):
    resumen_datos.append({
        'Base de Datos': nombre_tienda,
        'Filas': df.shape[0],
        'Columnas': df.shape[1]
    })

df_dimensiones = pd.DataFrame(resumen_datos)
print("RESUMEN DE DIMENSIONES")
display(df_dimensiones)

RESUMEN DE DIMENSIONES


Base de Datos  Filas  Columnas
0        tienda   2359        12
1       tienda2   2359        12
2       tienda3   2359        12
3       tienda4   2358        12

####*Conclusión:* Si bien hay una fila de diferencia, la estructura es homogénea.

##2.- Verificar la homogeneidad en en los cabezados de columnas y en el orden en cada caso.

####Se compararán todas las tiendas contra la primera, que será la referencial.

In [24]:
columnas_referencia = list(bases[0].columns)

print("VALIDACIÓN DE ESTRUCTURA DE COLUMNAS")
for nombre_tienda, df in zip(las_tiendas, bases):
    if list(df.columns) == columnas_referencia:
        print(f"{nombre_tienda.upper()}: Estructura coincidente.")
    else:
        diferencia = set(columnas_referencia) ^ set(df.columns)
        print(f"{nombre_tienda.upper()}: Discrepancia detectada. Diferencia en: {diferencia}")

VALIDACIÓN DE ESTRUCTURA DE COLUMNAS
TIENDA: Estructura coincidente.
TIENDA2: Estructura coincidente.
TIENDA3: Estructura coincidente.
TIENDA4: Estructura coincidente.


####*Conclusión:* Estructuralmente son coincidentes en ambos aspectos.

##3.- Verificación de la Calidad de los Datos en las bases.

####Se verá la existencia de datos nulos y en porcentajes; los tipos de datos, máximos y mínimos por cada tienda y valores únicos. Todo resumido.

In [25]:
for nombre_tienda, df in zip(las_tiendas, bases):
    # 1. Cálculos de calidad
    nulos = df.isnull().sum()
    porcentaje_nulos = (nulos / len(df)) * 100
    tipos_datos = df.dtypes
    unicos = df.nunique()

    # 2. Construcción del reporte
    reporte = pd.DataFrame({
        'Columna': nulos.index,
        'Tipo de Dato': tipos_datos.values,
        'Valores Nulos': nulos.values,
        '% Nulos': porcentaje_nulos.values,
        'Valores Únicos': unicos.values
    })
    reporte['% Nulos'] = reporte['% Nulos'].map('{:.2f}%'.format)

    # 3. Salida de Resultados
    print(f"REPORTE CRÍTICO DE DATOS: {nombre_tienda.upper()}")
    print(f"Estado: {df.shape[0]} filas | {df.duplicated().sum()} duplicados")
    display(reporte)

    # Resumen numérico
    columnas_numericas = df.select_dtypes(include=['number'])
    if not columnas_numericas.empty:
        print(f"Límites numéricos detectados:")
        display(columnas_numericas.describe().loc[['min', 'max']])

    print("\n" + "." * 30 + "\n")

REPORTE CRÍTICO DE DATOS: TIENDA
Estado: 2359 filas | 0 duplicados


Columna Tipo de Dato  Valores Nulos % Nulos  Valores Únicos
0                 Producto       object              0   0.00%              51
1   Categoría del Producto       object              0   0.00%               8
2                   Precio      float64              0   0.00%            1937
3           Costo de envío      float64              0   0.00%             753
4          Fecha de Compra       object              0   0.00%            1018
5                 Vendedor       object              0   0.00%              14
6          Lugar de Compra       object              0   0.00%              19
7             Calificación        int64              0   0.00%               5
8           Método de pago       object              0   0.00%               4
9       Cantidad de cuotas        int64              0   0.00%              15
10                     lat      float64              0   0.00%              19
11                     lon      float64              0   0.00%              19

Límites numéricos detectados:


Precio  Costo de envío  Calificación  Cantidad de cuotas       lat  \
min     7600.0             0.0           1.0                 1.0  -4.21528   
max  2977000.0        160800.0           5.0                24.0  11.54444   

          lon  
min -77.28111  
max -67.92390


..............................

REPORTE CRÍTICO DE DATOS: TIENDA2
Estado: 2359 filas | 0 duplicados


Columna Tipo de Dato  Valores Nulos % Nulos  Valores Únicos
0                 Producto       object              0   0.00%              51
1   Categoría del Producto       object              0   0.00%               8
2                   Precio      float64              0   0.00%            1908
3           Costo de envío      float64              0   0.00%             736
4          Fecha de Compra       object              0   0.00%            1025
5                 Vendedor       object              0   0.00%              14
6          Lugar de Compra       object              0   0.00%              19
7             Calificación        int64              0   0.00%               5
8           Método de pago       object              0   0.00%               4
9       Cantidad de cuotas        int64              0   0.00%              17
10                     lat      float64              0   0.00%              19
11                     lon      float64              0   0.00%              19

Límites numéricos detectados:


Precio  Costo de envío  Calificación  Cantidad de cuotas       lat  \
min     7600.0             0.0           1.0                 1.0  -4.21528   
max  2953400.0        162500.0           5.0                24.0  11.54444   

          lon  
min -77.28111  
max -67.92390


..............................

REPORTE CRÍTICO DE DATOS: TIENDA3
Estado: 2359 filas | 0 duplicados


Columna Tipo de Dato  Valores Nulos % Nulos  Valores Únicos
0                 Producto       object              0   0.00%              51
1   Categoría del Producto       object              0   0.00%               8
2                   Precio      float64              0   0.00%            1946
3           Costo de envío      float64              0   0.00%             737
4          Fecha de Compra       object              0   0.00%            1024
5                 Vendedor       object              0   0.00%              14
6          Lugar de Compra       object              0   0.00%              19
7             Calificación        int64              0   0.00%               5
8           Método de pago       object              0   0.00%               4
9       Cantidad de cuotas        int64              0   0.00%              17
10                     lat      float64              0   0.00%              19
11                     lon      float64              0   0.00%              19

Límites numéricos detectados:


Precio  Costo de envío  Calificación  Cantidad de cuotas       lat  \
min     7600.0             0.0           1.0                 1.0  -4.21528   
max  2968200.0        158400.0           5.0                24.0  11.54444   

          lon  
min -77.28111  
max -67.92390


..............................

REPORTE CRÍTICO DE DATOS: TIENDA4
Estado: 2358 filas | 0 duplicados


Columna Tipo de Dato  Valores Nulos % Nulos  Valores Únicos
0                 Producto       object              0   0.00%              51
1   Categoría del Producto       object              0   0.00%               8
2                   Precio      float64              0   0.00%            1867
3           Costo de envío      float64              0   0.00%             692
4          Fecha de Compra       object              0   0.00%            1030
5                 Vendedor       object              0   0.00%              14
6          Lugar de Compra       object              0   0.00%              19
7             Calificación        int64              0   0.00%               5
8           Método de pago       object              0   0.00%               4
9       Cantidad de cuotas        int64              0   0.00%              13
10                     lat      float64              0   0.00%              19
11                     lon      float64              0   0.00%              19

Límites numéricos detectados:


Precio  Costo de envío  Calificación  Cantidad de cuotas       lat  \
min     7600.0             0.0           1.0                 1.0  -4.21528   
max  2902200.0        154700.0           5.0                18.0  11.54444   

          lon  
min -77.28111  
max -67.92390


..............................



####*Conclusión:* Hasta aquí, tenemos  buena calidad de datos. El único detalle es el **tipo de dato de la fecha**, es object, lo que pudiera complicar dependiendo el análisi solicitado. Se deja en veremos.

##4.-Valores Blanco o cero

In [26]:
for nombre_tienda, df in zip(las_tiendas, bases):
    # 1. Preparación de datos
    precios_cero = (df['Precio'] <= 0).sum()
    ciudades_unicas = df['Lugar de Compra'].unique()
    fechas_temp = pd.to_datetime(df['Fecha de Compra'], dayfirst=True, errors='coerce')

    # 2. Diccionario de hallazgos
    diagnostico = {
        "Métrica": [
            "Precios Inválidos (<= 0)",
            "Total Ciudades Distintas",
            "Fecha Más Antigua",
            "Fecha Más Reciente",
            "¿Tiene Duplicados?"
        ],
        "Resultado": [
            precios_cero,
            len(ciudades_unicas),
            fechas_temp.min().strftime('%d/%m/%Y') if not fechas_temp.dropna().empty else "N/A",
            fechas_temp.max().strftime('%d/%m/%Y') if not fechas_temp.dropna().empty else "N/A",
            "Sí" if df.duplicated().sum() > 0 else "No"
        ]
    }

    # 3. Salida visual
    print(f"ESTADO DE SALUD: {nombre_tienda.upper()}")
    display(pd.DataFrame(diagnostico))
    print("\n")

ESTADO DE SALUD: TIENDA


Métrica   Resultado
0  Precios Inválidos (<= 0)           0
1  Total Ciudades Distintas          19
2         Fecha Más Antigua  01/01/2020
3        Fecha Más Reciente  31/03/2023
4        ¿Tiene Duplicados?          No



ESTADO DE SALUD: TIENDA2


Métrica   Resultado
0  Precios Inválidos (<= 0)           0
1  Total Ciudades Distintas          19
2         Fecha Más Antigua  01/01/2020
3        Fecha Más Reciente  31/03/2023
4        ¿Tiene Duplicados?          No



ESTADO DE SALUD: TIENDA3


Métrica   Resultado
0  Precios Inválidos (<= 0)           0
1  Total Ciudades Distintas          19
2         Fecha Más Antigua  01/01/2020
3        Fecha Más Reciente  31/03/2023
4        ¿Tiene Duplicados?          No



ESTADO DE SALUD: TIENDA4


Métrica   Resultado
0  Precios Inválidos (<= 0)           0
1  Total Ciudades Distintas          19
2         Fecha Más Antigua  01/01/2020
3        Fecha Más Reciente  30/03/2023
4        ¿Tiene Duplicados?          No

####*Conclusión local:* en ésgte analisis los resultados son favorables al análisis.

# CONCLUSIÓN:
1. **Consistencia Temporal Perfecta**.
Las tiendas 1, 2 y 3 tienen exactamente el mismo rango (hasta el 31/03/2023).La Tienda 4 termina el 30/03/2023. Esto explica perfectamente por qué esa base tiene 2,359 filas en lugar de las 2,360 de las demás. No es un error, es simplemente que falta el último día de registros, lo cual es una discrepancia mínima y totalmente lógica.

2. **Integridad de Precios**.
0 Precios Inválidos: Esto es crítico. Significa que no hubo errores de captura, devoluciones mal registradas o valores nulos en la columna más importante para el negocio (el ingreso).

3. **Coincidencia Geográfica**.
El hecho de que todas tengan 19 ciudades distintas, sugiere que la estructura de operación es uniforme en todas las sucursales. Esto facilitará muchísimo las comparaciones entre tiendas, ya que "juegan bajo las mismas reglas" geográficas.

4. **Limpieza Absoluta**
Sin duplicados: No hay que realizar la limpieza de datos. Se puede confiar en que cada fila es una transacción única y real.

#1. Análisis de facturación



# 2. Ventas por categoría

# 3. Calificación promedio de la tienda


# 4. Productos más y menos vendidos

# 5. Envío promedio por tienda